### PR TITLE
Feature/Removing url parameter max length

### PR DESCRIPTION
**Pull request by SNiukalov**
Fixes [#2702](https://github.com/smartdevicelink/sdl_core/issues/2702)

This PR is **ready** for review.

### Risk
This PR makes **minor** API changes.

### Testing Plan
Unit tests and ATF scripts.

Use scripts from:
[smartdevicelink/sdl_atf_test_scripts#2353](https://github.com/smartdevicelink/sdl_atf_test_scripts/pull/2353)

### Summary
PR contains sdl_core implementation of [SDL 0200] Removing URL Parameter Max Length

### NOTE 
Please, don't **`auto merge`** this pull request, because it contains changes for the rpc_spec, which should be removed before merge. 
These changes are temporary and required for successful build & testing. 
SDL Core has dependency from other repository: https://github.com/smartdevicelink/rpc_spec
Contributor has no rights at the moment to create branches in the repository specified above. 

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

**Comment by SNiukalov** [comment]
Leave commetns for remove rpc_spec submodule changes

File: src/components/application_manager/src/message_helper/message_helper.cc [hunk 1]
```diff
@@ -2357,25 +2357,22 @@ void MessageHelper::SendNaviStartStream(const int32_t app_id,
...
+    url += std::to_string(port);
   } else if ("pipe" == app_mngr.get_settings().video_server_type()) {
-    snprintf(url,
-             sizeof(url) / sizeof(url[0]),
-             "%s",
-             app_mngr.get_settings().named_video_pipe_path().c_str());
+    url.reserve(PATH_MAX);
+    url.insert(
+        0, app_mngr.get_settings().named_video_pipe_path(), 0, PATH_MAX - 1);
+    DCHECK(PATH_MAX > url.length());
```
Comment (SNiukalov) [inline src/components/application_manager/src/message_helper/message_helper.cc]: remove DCHECK

File: src/components/application_manager/src/message_helper/message_helper.cc [hunk 2]
```diff
@@ -2357,25 +2357,22 @@ void MessageHelper::SendNaviStartStream(const int32_t app_id,
...
   } else {
-    int snprintf_result =
-        snprintf(url,
-                 sizeof(url) / sizeof(url[0]),
-                 "%s",
-                 app_mngr.get_settings().video_stream_file().c_str());
-    DCHECK(snprintf_result);
+    url.reserve(PATH_MAX);
+    url.insert(0, app_mngr.get_settings().video_stream_file(), 0, PATH_MAX - 1);
+    DCHECK(PATH_MAX > url.length());
```
Comment (SNiukalov) [inline src/components/application_manager/src/message_helper/message_helper.cc]: remove DCHECK

File: src/components/application_manager/src/message_helper/message_helper.cc [hunk 3]
```diff
@@ -2414,25 +2411,22 @@ void MessageHelper::SendAudioStartStream(const int32_t app_id,
...
+    url += std::to_string(port);
   } else if ("pipe" == app_mngr.get_settings().audio_server_type()) {
-    snprintf(url,
-             sizeof(url) / sizeof(url[0]),
-             "%s",
-             app_mngr.get_settings().named_audio_pipe_path().c_str());
+    url.reserve(PATH_MAX);
+    url.insert(
+        0, app_mngr.get_settings().named_audio_pipe_path(), 0, PATH_MAX - 1);
+    DCHECK(PATH_MAX > url.length());
```
Comment (SNiukalov) [inline src/components/application_manager/src/message_helper/message_helper.cc]: remove DCHECK

File: src/components/application_manager/src/message_helper/message_helper.cc [hunk 4]
```diff
@@ -2414,25 +2411,22 @@ void MessageHelper::SendAudioStartStream(const int32_t app_id,
...
   } else {
-    int snprintf_result =
-        snprintf(url,
-                 sizeof(url) / sizeof(url[0]),
-                 "%s",
-                 app_mngr.get_settings().audio_stream_file().c_str());
-    DCHECK(snprintf_result);
+    url.reserve(PATH_MAX);
+    url.insert(0, app_mngr.get_settings().audio_stream_file(), 0, PATH_MAX - 1);
+    DCHECK(PATH_MAX > url.length());
```
Comment (SNiukalov) [inline src/components/application_manager/src/message_helper/message_helper.cc]: remove DCHECK

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/test/commands/mobile/on_system_request_notification_test.cc [hunk 5]
```diff
@@ -160,6 +175,57 @@ TEST_F(OnSystemRequestNotificationTest, Run_HTTPType_SUCCESS) {
             (*msg)[strings::params][strings::protocol_version].asInt());
 }
```
Comment (SNiukalov) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/test/commands/mobile/on_system_request_notification_test.cc]: Stop review

File: .gitmodules [hunk 6]
```diff
@@ -3,4 +3,5 @@
 	url = https://github.com/open-source-parsers/jsoncpp.git
 [submodule "tools/rpc_spec"]
 	path = tools/rpc_spec
-	url = https://github.com/smartdevicelink/rpc_spec.git
+	url = https://github.com/SNiukalov/rpc_spec.git
```
Comment (AByzhynar) [inline .gitmodules]: @SNiukalov Why do we have your personal repo here?

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/test/commands/hmi/navi_audio_start_stream_request_test.cc [hunk 7]
```diff
@@ -74,11 +75,24 @@ class AudioStartStreamRequestTest
...
+        am::MessageType::kRequest;
+    (*message)[am::strings::params][am::strings::function_id] =
+        static_cast<int32_t>(hmi_apis::FunctionID::Navigation_StartAudioStream);
+    (*message)[am::strings::msg_params][am::strings::app_id] = kHmiAppId;
+  }
+
   std::pair<uint32_t, int32_t> start_stream_retry_amount_;
   MessageSharedPtr msg_;
   std::shared_ptr<AudioStartStreamRequest> command_;
+  static const std::string big_url_;
```
Comment (AByzhynar) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/test/commands/hmi/navi_audio_start_stream_request_test.cc]: @SNiukalov Why did you make it static?

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/test/commands/mobile/on_system_request_notification_test.cc (hunk 5 above)
Comment (AByzhynar) [reply src/components/application_manager/rpc_plugins/sdl_rpc_plugin/test/commands/mobile/on_system_request_notification_test.cc]: @SNiukalov What does it mean `Stop review` ?

File: src/components/application_manager/src/message_helper/message_helper.cc (hunk 1 above)
Comment (SNiukalov) [reply src/components/application_manager/src/message_helper/message_helper.cc]: Fixed in:https://github.com/smartdevicelink/sdl_core/pull/3259/commits/d07a4abca1bb8d88ff002ffa1ddd72b725d733c9

File: src/components/application_manager/src/message_helper/message_helper.cc (hunk 2 above)
Comment (SNiukalov) [reply src/components/application_manager/src/message_helper/message_helper.cc]: Fixed in:https://github.com/smartdevicelink/sdl_core/pull/3259/commits/d07a4abca1bb8d88ff002ffa1ddd72b725d733c9

File: src/components/application_manager/src/message_helper/message_helper.cc (hunk 3 above)
Comment (SNiukalov) [reply src/components/application_manager/src/message_helper/message_helper.cc]: Fixed in:https://github.com/smartdevicelink/sdl_core/pull/3259/commits/d07a4abca1bb8d88ff002ffa1ddd72b725d733c9

File: src/components/application_manager/src/message_helper/message_helper.cc (hunk 4 above)
Comment (SNiukalov) [reply src/components/application_manager/src/message_helper/message_helper.cc]: Fixed in:https://github.com/smartdevicelink/sdl_core/pull/3259/commits/d07a4abca1bb8d88ff002ffa1ddd72b725d733c9

File: .gitmodules (hunk 6 above)
Comment (SNiukalov) [reply .gitmodules]: @AByzhynar 
Due to the [latest changes committed by the PM](https://github.com/smartdevicelink/sdl_core/pull/3270/files),
currently to build SDL it is necessary to use MOBILE_API from the https://github.com/smartdevicelink/rpc_spec.

Luxoft team does not have access to create branches in https://github.com/smartdevicelink/rpc_spec repo, but MOBILE_API is used by unit tests.

Therefore, changes to https://github.com/smartdevicelink/rpc_spec repo were created in SNiukalov:feature/removing_url_parameter_max_length branch

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/test/commands/hmi/navi_audio_start_stream_request_test.cc (hunk 7 above)
Comment (SNiukalov) [reply src/components/application_manager/rpc_plugins/sdl_rpc_plugin/test/commands/hmi/navi_audio_start_stream_request_test.cc]: @AByzhynar 
Made it static for performance optimization.

File: .gitmodules (hunk 6 above)
Comment (AByzhynar) [reply .gitmodules]: @SNiukalov Clear.Thank you.

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/test/commands/hmi/navi_audio_start_stream_request_test.cc (hunk 7 above)
Comment (AByzhynar) [reply src/components/application_manager/rpc_plugins/sdl_rpc_plugin/test/commands/hmi/navi_audio_start_stream_request_test.cc]: @SNiukalov If I understand optimization purpose in this case correctly - better to do it in C++ way:
```cpp
namespace {
const std::string kBigURL(20000u, 'a');
}
```

Comment (SNiukalov) [reply src/components/application_manager/rpc_plugins/sdl_rpc_plugin/test/commands/hmi/navi_audio_start_stream_request_test.cc]: @AByzhynar 
This is just alternative option

File: src/components/application_manager/test/message_helper/message_helper_test.cc [hunk 8]
```diff
@@ -359,10 +365,10 @@ TEST(MessageHelperTestCreate,
...
       .Times(AtLeast(5))
       .WillRepeatedly(Return(1u));
-  application_manager_test::MockApplicationManager mock_application_manager;
+  application_manager_test::MockApplicationManager mock_application_manager_;
   smart_objects::SmartObjectList ptr =
       MessageHelper::CreateAddVRCommandRequestFromChoiceToHMI(
-          appSharedMock, mock_application_manager);
+          appSharedMock, mock_application_manager_);
 
   EXPECT_FALSE(ptr.empty());
```
Comment (SNiukalov) [inline src/components/application_manager/test/message_helper/message_helper_test.cc]: In this case I had mistake and rename local variable according to code style for the class member variable.
This fixed in separate commit:https://github.com/smartdevicelink/sdl_core/pull/3259/commits/56a9fbc7389daa47d95b91e9f20807e33b20298a

**Comment by SNiukalov** [comment]
@yang1070 please approve these changes

File: .gitmodules [hunk 9]
```diff
@@ -3,4 +3,5 @@
 	url = https://github.com/open-source-parsers/jsoncpp.git
 [submodule "tools/rpc_spec"]
 	path = tools/rpc_spec
-	url = https://github.com/smartdevicelink/rpc_spec.git
```
Comment (yang1070) [inline .gitmodules]: this shall be removed before merge.

Comment (SNiukalov) [reply .gitmodules]: @yang1070 
Thank you for leaving a comment.
Please approve these changes.

**Comment by SNiukalov** [comment]
@theresalech PR is ready for Livio team review. Thank you!

**Comment by theresalech** [comment]
Hi @SNiukalov, thank you for letting us know!  This feature is not planned for the Core 6.1 release, but we will make note that the PR is ready for review if the Steering Committee chooses to include this feature in the next release of SDL Core, planned for later this year.

File: src/components/application_manager/src/message_helper/message_helper.cc [hunk 10]
```diff
@@ -2414,25 +2409,20 @@ void MessageHelper::SendAudioStartStream(const int32_t app_id,
...
+    url += ":";
+    url += std::to_string(port);
   } else if ("pipe" == app_mngr.get_settings().audio_server_type()) {
-    snprintf(url,
-             sizeof(url) / sizeof(url[0]),
-             "%s",
-             app_mngr.get_settings().named_audio_pipe_path().c_str());
+    url.reserve(PATH_MAX);
+    url.insert(
+        0, app_mngr.get_settings().named_audio_pipe_path(), 0, PATH_MAX - 1);
```
Comment (ShobhitAd) [inline src/components/application_manager/src/message_helper/message_helper.cc:2422]:
```suggestion
    url = std::string(app_mngr.get_settings().named_audio_pipe_path(), 0, PATH_MAX);
```
should also work.

Comment (SNiukalov) [reply src/components/application_manager/src/message_helper/message_helper.cc:2422]: @ShobhitAd 
Yes, but according to define for MAX_PATH - "chars in a path name including nul".
So we reserve 1 symbol for the null character

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/test/commands/hmi/navi_start_stream_request_test.cc [hunk 11]
```diff
@@ -113,6 +127,51 @@ TEST_F(NaviStartStreamRequestTest, Run_HmiInterfaceAvailable_SentRequest) {
...
+  ON_CALL(app_mngr_, application(kHmiAppId)).WillByDefault(Return(mock_app));
+  ON_CALL(*mock_app, hmi_app_id()).WillByDefault(Return(kHmiAppId));
+
+  EXPECT_CALL(*mock_app, set_video_streaming_allowed(true));
+  EXPECT_CALL(mock_rpc_service_, SendMessageToHMI(msg_));
+
+  ASSERT_TRUE(command_->Init());
+  command_->Run();
+
+  EXPECT_EQ(big_url_,
```
[the remark this answers is not in the document]
Comment (SNiukalov) [reply src/components/application_manager/rpc_plugins/sdl_rpc_plugin/test/commands/hmi/navi_start_stream_request_test.cc:150]: @ShobhitAd 
Yes, you right, this check is not required fixed in:https://github.com/smartdevicelink/sdl_core/commit/29aa4abe5f7af44f50bd4e306bec78178279abcc

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/test/commands/hmi/navi_audio_start_stream_request_test.cc [hunk 12]
```diff
@@ -114,6 +128,51 @@ TEST_F(AudioStartStreamRequestTest, Run_HmiInterfaceAvailable_SentRequest) {
...
+  ON_CALL(app_mngr_, application(kHmiAppId)).WillByDefault(Return(mock_app));
+  ON_CALL(*mock_app, hmi_app_id()).WillByDefault(Return(kHmiAppId));
+
+  EXPECT_CALL(*mock_app, set_audio_streaming_allowed(true));
+  EXPECT_CALL(mock_rpc_service_, SendMessageToHMI(msg_));
+
+  ASSERT_TRUE(command_->Init());
+  command_->Run();
+
+  EXPECT_EQ(big_url_,
```
[the remark this answers is not in the document]
Comment (SNiukalov) [reply src/components/application_manager/rpc_plugins/sdl_rpc_plugin/test/commands/hmi/navi_audio_start_stream_request_test.cc:151]: @ShobhitAd 
Yes, you right, this check is not required fixed in:https://github.com/smartdevicelink/sdl_core/commit/29aa4abe5f7af44f50bd4e306bec78178279abcc

**Comment by LuxoftAKutsan** [comment]
@ShobhitAd should we track the status of smartdevicelink/rpc_spec#225 to update this PR? And then after the update you will merge it?

**Comment by ShobhitAd** [comment]
The RPC spec PR for this feature https://github.com/smartdevicelink/rpc_spec/pull/225 has been merged into develop. Please update the `.gitmodules` file to use the `smartdevicelink` remote and update the `tools/rpc_spec` submodule commit.

**Comment by LitvinenkoIra** [comment]
@ShobhitAd Thank you, please check 8e64f58

**Pull request merged**